### DISCUSSION TITLE
What does --variable-width-glyphs do now?

### DISCUSSION THREAD

**Discussion by nathanielevan** [ryanoasis/nerd-fonts#1103] [General]
Hi! I understand that in the latest release all glyphs are made to be either single-width or double-width; if I remember correctly `--variable-width-glyphs`  was supposed to let glyphs take its full width or something(?), so what does the option do now? Just asking for some clarity on the matter, thanks in advance!

**Comment by Finii** [comment]
Thanks for the question. You are right this is a bit shady and an all-encompassing explanation is nowhere in sight.

Let me first explain the current (or near future) state and afterwards some historical information.
Some thoughts can also be gleamed from #969, and there are a lot of threads scattered about regarding naming.

## Nerd Fonts variants

Depending on what you want to do the added symbols needed to be patched-in in different ways. At the moment there are three major different flavors:

* **`Nerd Font Mono`** (a strictly monospaced variant, created with `--mono`)
* **`Nerd Font`** (a somehow monospaced variant, maybe)
* **`Nerd Font Propo`** (a not monospaced variant, created with `--variable-width-glyphs`)

This all revolves arount 'monospaced'. What is a 'monospaced' font? One where all characters (font people call them glyphs) are exactly the same width. Like on an old style typewrite (if you are that old). The characters form some kind of grid, and regardless of _which_ letters you type, on consecutive lines the 23rd letter is exactly above the other 23rd letter. In this example the upper text is set in a monospaced font, the lower in a not monospaced font:

![image](https://user-images.githubusercontent.com/16012374/216336263-507e9502-f961-4601-8c4b-eb43b5dafd3b.png)

_Note that all letters line up nicely on the pink line with the monospaced font, but not for the non-monospaced one. Also note the different relative length of line 2 and 3, where line 3 has more 'narrow' letters like `i` in it._

Having a monospaced font is something a lot programmers like, because multi-line code often looks better this way.

So what are the Nerd Font flavors' differences and use-cases?

### Nerd Font Mono

These variant is only available when the source font is monospaced.
Here all characters have the same width (one 'cell' width), and the patched in symbols are scaled down/up that they fit exactly in one cell.

![image](https://user-images.githubusercontent.com/16012374/216339510-5791cbae-716e-4971-9774-cda2a636c9c1.png)

In the example above two characters (one `e` and one `,`) have been replaced with `nf-cod-github`, the github logo.
* the 'grid of letters' is preserved
* the following letters are easy to see (`greet`) because the symbol is as small as all letters
* the symbol is rather small, looks a bit too small (but is indeed maximized in one 'cell')

**The font is for programmers or other people that want a fixed and dependable grit of letters.**
Also the symbols fit into the grid, but the price is that they are small.

### Nerd Font

Icons too small? We can scale the icons to a bigger size (limited to 2 cells wide (and 1 cell high)) with this font.
But we want also to preserve the 'grid'. The only way to achieve this is to have the symbols extend out of the cell into the next, but keep the 'advance width' at 'one cell'.

![image](https://user-images.githubusercontent.com/16012374/216343767-dc1cec67-0812-4b02-8992-c2ecb95403b5.png)

As you see this only looks good if the 'cell' after the symbol is empty.
This is supported by a lot tools like `exa` and Powerline prompts and whatever.

* big symbols
* grind unconditionally preserved
* needs blanks after symbols

Note that they do not 'occupy' the next cell. In principle we could also create a font were the symbols are '2 cells wide'. But that would not be detected as 'monospaced' anymore. It will break your carefully vertically aligned tables when a 1 cell glyph is replaced by a 2 cell glyph.

**This font is for programmers and other people that want a grid and big symbols, and can accept strange overhanging glyphs**

### Nerd Font Propo

But sometimes you are not bound by any 'grid'. You edit slides for a talk or want to use symbols in GUI contexts. Here you want bigger symbols but the 'advance width' should match the visual width of the symbol. The font has many different widths, matching the individually different symbols and is thus not 'monospaced' but a proportional font (that is the opposite of 'monospaced')

![image](https://user-images.githubusercontent.com/16012374/216347282-5d783b66-cac0-498b-bd69-0bd63016460d.png)

_It looks like the grid is preserved, at least when looking at line 2 and 3, but line 4 shows that they just both break the grid with the same amount._

* big symbols
* no blanks after symbols needed
* grid destroyed

**This font is for usage in proportional contexts like text documents or GUI text elements**

### History

#### up to and including `v2.0.0`
In the beginning Nerd Fonts were all `Nerd Font Mono` like fonts and there were no other choices. Then the `Nerd Font` variant had been added for increased symbol readability.

#### This was `v2.1.0` only
Then there were complaints from people who use Nerd Fonts for GUIs and we had `Nerd Font Mono` and `Nerd Font Propo` which was called `Nerd Font` - the `Nerd Font` variant mentioned above vanished.

#### Version `v2.2.0` and afterwards
That made a lot of terminal users unhappy and the change was rolled back. So now the prepatched variants are `Nerd Font` and `Nerd Font Mono`. It is planned to make the `Nerd Font Propo` part of the prepatched release with `v3.0.0`.

_Edit: Add forgotten history paragraph_

**Reply by Finii** [reply]
The `Propo` part in the name is still missing, as are some other details with the naming not finalized yet.
The font names have some issues that can only be resolved by renaming them, which will break existing documents, so that has to be considered very carefully. If we get it wrong (again) we would need another name-breaking change and noone wants that to happen.

So the `Propo` part is #1028, but that needs some work still.

**Reply by nathanielevan** [reply]
Right, that clears things up, thank you! So I understand that the "regular" Nerd Font variant comes with icons at most 2 cells wide and they're allowed to overhang -- will the glyphs in Propo also be at most 2 cells wide, or can they be as wide as they need to as long as they don't exceed the height of the cell? There are icons such as `nf-fa-battery` (U+F240 / ) that are significantly wider than it is tall, and I think it's OK to let them take their full width in Propo, but I'm willing to accept if you have different plans for it.

**Reply by Finii** [reply]
The scaling in `Nerd Font` and `Nerd Font Propo` is the same, just the advance width is different:

While in `Propo` all 'original' width are kept are the added symbols in `Nerd Font` set to have one 'monospace' width regardless of visual with. And why do I mention that specifically here: We can also non-monospace original fonts (like `Ubuntu` vs `Ubuntu Mono`), that have proportional style letters. For these we 'invent' a monospace-cell-size just for the symbols added.

And then, for `Nerd Font` and `Nerd Font Propo` the symbols are scaled **down** to be maximally 2 cells wide. But they are not scaled **up**. So typically they will be ~1.6-1.8 width wide, depending onb their original definition and to-be-patched-font width.

> There are icons such as nf-fa-battery ... I think it's OK to let them take their full width in Propo

![image](https://user-images.githubusercontent.com/16012374/216531089-08da5a2a-81d2-419d-9e10-bde01c2206e3.png)

You are probably right.

![image](https://user-images.githubusercontent.com/16012374/216532061-a51cdb26-729b-47d2-addd-147bffc3bfa4.png)

_Width 1200 vs 2400_

Without the 2cell limit:

![image](https://user-images.githubusercontent.com/16012374/216533304-90aad69d-9f5e-42d4-93b5-08010e31a903.png)

_Width 1200 vs 2633_

**Reply by Finii** [reply]
Another details ....

The 2-cell thing is disabled with very wide and stubby fonts, because it looks really strange, esp the powerline glyphs

```python
        # For very wide (almost square or wider) fonts we do not want to generate 2 cell wide Powerline glyphs
          if self.font_dim['height'] * 1.8 < self.font_dim['width'] * 2:
              print("Very wide and short font, disabling 2 cell Powerline glyphs")
              self.font_extrawide = True
```

I'm unsure if that `extrawide` is used for all, Powerline and other glyphs, or only Powerline :thinking:

_Edit:_

Ah, it's just the Powerline glyphs :sweat_smile: ... lucky me :wink: 

![image](https://user-images.githubusercontent.com/16012374/216534991-0bf11adf-4545-41eb-b2c1-88b0b93fdf54.png)

_Patched with `extrawide` detection artificially always firing_

**Reply by ChristopherHaws** [reply]
@Finii Can you explain what the difference between `Nerd Font` and `NF` variants are? Is `NF` another term for `Propo`?

![image](https://user-images.githubusercontent.com/5934505/227826189-dbfe85b7-21bd-4aac-a157-08d14a42cfce.png)

**Reply by Finii** [reply]
That is just an abbreviation:

`Nerd Font` -> `NF`
`Nerd Font Mono` -> `NFM`
`Nerd Font Propo` -> `NFP`

The short variants are the so-called `Windows Compatible` fonts; which is not a good name as all fonts are working on Windows, just some very old Microsoft applications like `Word for Mac` (on Mac not Windows) need the shorter names.

See also
* #969
* #891

**Reply by RogueScholar** [reply]
Absolutely flawless illumination of this topic, @Finii. How do we get this into the README? I'd love to see it get dropped in pretty much verbatim, honestly. This is foundational knowledge about the structure of the project itself and since we don't have an installer GUI to present it in, it really should be front and center for users before they begin downloading and installing things themselves.
